### ISSUE TITLE
Unit test for interdomain modify delete/add failing

### ISSUE THREAD

**Issue by ibaldin** [RENCI-NRIG/orca5#180]
We are using branch 178_nodegroup_interface_sameas for this. There are frequent (however not 100% of the time) failures of two unit tests:

 - OrcaXmlrpcHandlerTest.testNodeWithTwoInterfacesInterdomainDeleteAdd()
 - OrcaXmlrpcHandlerTest.testMixedDomainMultiStepModify()

It appears to be caused by improperly set num.parent.new property on the modified reservation. In the valid case its value is 1, in invalid case its value is null. 

Here are example properties for valid and invalid cases:

```
hostname Node0 had localProperties count 30 (aad0d999-feb5-4366-8f78-16461bee201e)
{unit.number.storage=0,
 unit.eth2.ip=172.16.0.2/30,
 unit.eth2.mac=fe:16:3e:00:63:71,
 unit.eth2.parent.url=http://geni-orca.renci.org/owl/rciNet.rdf#rciNet/Domain/vlan/71fced19-b9b5-4d4b-8f17-48e806108bee/vlan,
 num.parent.new=1,
 unit.eth2.uuid=c2e881d5-6cd2-43b9-8592-75891299e80d,
 unit.eth3.netmask=255.255.255.252,
 parent.exist.1=b4ea648f-fa02-4746-bb71-f24f6b489bc5,
 unit.eth1.netmask=255.255.255.252,
 parent.exist.0=1d6eaad3-81ce-42d8-9f4a-25946bb249e3,
 unit.eth3.ip=172.16.0.10/30,
 unit.eth3.parent.url=http://geni-orca.renci.org/owl/rciNet.rdf#rciNet/Domain/vlan/02dcf90e-b96f-4181-aebc-95f862084697/vlan,
 num.parent.exist=2,
 unit.url=http://geni-orca.renci.org/owl/2e33b59b-0e3e-4b40-a64d-d9f3c055326f#Node0,
 xmlrpc.user.dn=test,
 unit.eth1.hosteth=vlan-data,
 unit.eth1.mac=fe:16:3e:00:25:6a,
 unit.number.interface=3,
 unit.eth2.netmask=255.255.255.252,
 unit.eth3.mac=fe:16:3e:00:12:79,
 unit.slice.name=testNodeWithTwoInterfacesInterdomainDeleteAdd,
 local.isVM=1,
 unit.eth3.hosteth=vlan-data,
 unit.eth1.ip=172.16.0.5/30,
 unit.hostname=Node0,
 unit.eth2.hosteth=vlan-data,
 parent.new.0=a04ec07d-07b2-4e73-9ce3-f976a4988c1f,
 unit.eth1.parent.url=http://geni-orca.renci.org/owl/rciNet.rdf#rciNet/Domain/vlan/8482ad82-3f28-47bc-9405-4424b6f1f4f8/vlan,
 modify.version=2,
 unit.eth1.uuid=cda5672d-e0f1-432b-8287-49f7724a4b47
}

hostname Node0 had localProperties count 23 (98e6a2e9-bb3e-4c4d-826d-cbb8e248a685)
{unit.number.storage=0,
 unit.eth2.ip=172.16.0.2/30,
 unit.eth2.mac=fe:16:3e:00:45:4b,
 unit.eth2.parent.url=http://geni-orca.renci.org/owl/rciNet.rdf#rciNet/Domain/vlan/35d35f8a-890a-4f88-a34e-3e6d9ee20611/vlan,
 unit.eth2.uuid=12c076f2-09b8-471d-a84a-c4a1d781818d,
 parent.exist.1=d1c25a56-645f-4821-bea4-95b02e2d594c,
 unit.eth1.netmask=255.255.255.252,
 parent.exist.0=8f14a30c-727e-4f6f-a0f2-87ec50697ee6,
 num.parent.exist=2,
 unit.url=http://geni-orca.renci.org/owl/2e33b59b-0e3e-4b40-a64d-d9f3c055326f#Node0,
 xmlrpc.user.dn=test,
 unit.eth1.hosteth=vlan-data,
 unit.eth1.mac=fe:16:3e:00:14:06,
 unit.number.interface=2,
 unit.eth2.netmask=255.255.255.252,
 unit.slice.name=testNodeWithTwoInterfacesInterdomainDeleteAdd,
 local.isVM=1,
 unit.eth1.ip=172.16.0.5/30,
 unit.hostname=Node0,
 unit.eth2.hosteth=vlan-data,
 unit.eth1.parent.url=http://geni-orca.renci.org/owl/rciNet.rdf#rciNet/Domain/vlan/5eb81703-e857-42d0-a11d-656ca6a4e597/vlan,
 modify.version=2,
 unit.eth1.uuid=5a9713a2-6183-478f-ace9-0a2feaac1762
}
```

The log messages for valid and invalid cases also differ - there is an additional reservation being added to status watch:

```
VALID 

2018-01-04 12:33:38,852 [main] DEBUG controller.OrcaXmlrpcHandler-(OrcaXmlrpcHandler.java:944) - Issuing modify demand for reservation: e2b5cfd0-eefc-4c1f-a059-cd8324c2bb97
2018-01-04 12:33:38,854 [main] DEBUG controller.OrcaXmlrpcHandler-(OrcaXmlrpcHandler.java:969) - addActiveStatuWatch:numExistParent=2
2018-01-04 12:33:38,854 [main] DEBUG controller.OrcaXmlrpcHandler-(OrcaXmlrpcHandler.java:982) - addActiveStatuWatch:numNewParent=1
2018-01-04 12:33:38,854 [main] DEBUG controller.OrcaXmlrpcHandler-(OrcaXmlrpcHandler.java:992) - addActiveStatuWatch:[6fb87426-02f8-4ccf-aaef-747cb48a2d29, 43648d72-602b-469e-9832-b2542e0363a3, 9ae1929d-1b20-4fea-b8cb-6f0bb8e6352e];;self=[e2b5cfd0-eefc-4c1f-a059-cd8324c2bb97]

INVALID 

2018-01-04 13:13:44,906 [main] DEBUG controller.OrcaXmlrpcHandler-(OrcaXmlrpcHandler.java:944) - Issuing modify demand for reservation: fbdaad0a-072b-40db-adb7-77b3f295ade5
2018-01-04 13:13:44,908 [main] DEBUG controller.OrcaXmlrpcHandler-(OrcaXmlrpcHandler.java:969) - addActiveStatuWatch:numExistParent=2
2018-01-04 13:13:44,908 [main] DEBUG controller.OrcaXmlrpcHandler-(OrcaXmlrpcHandler.java:982) - addActiveStatuWatch:numNewParent=null
2018-01-04 13:13:44,908 [main] DEBUG controller.OrcaXmlrpcHandler-(OrcaXmlrpcHandler.java:992) - addActiveStatuWatch:[10712c2c-26df-4e2d-8156-20da5ff6ad11, a0450abe-2793-41cf-9356-ac4619ac04b3];;self=[fbdaad0a-072b-40db-adb7-77b3f295ade5]
```

The initial thought that this is somehow being caused by a race condition for a inter-domain delayed reservation callback does not appear to be correct. Setting a delay between consecutive modify operations of 1min didn't fix the problem. 

Haven't tried delaying the first modify after create yet.

**Comment by ibaldin** [comment]
The problem appears to be that in ReservationConverter.modifyReservations() there is a loop
```
for (Entry<DomainElement, OntResource> parent : dd.getPrecededBySet())
```
that iterates over preceded by entry set of domain elements-URI mappings. When the test fails the set contains one fewer URIs - it appears the missing one should be there to help set num.parent.new property:
```
m_r_map for uri http://geni-orca.renci.org/owl/rciNet.rdf#rciNet/Domain/vlan/40aa9274-83a5-47a2-989d-65a1988863d8/vlan returned null
contains test false

m_r_map for uri http://geni-orca.renci.org/owl/rciNet.rdf#rciNet/Domain/vlan/831f620a-f764-4c2b-88b8-c086682b5c19/vlan returned null
contains test false

m_r_map for uri http://geni-orca.renci.org/owl/rciNet.rdf#rciNet/Domain/vlan/7630992e-ac99-4a78-beee-1b504b29cef0/vlan returned orca.manage.beans.LeaseReservationMng@5c65d200
contains test false

ModifiedReservation Parent new:http://geni-orca.renci.org/owl/rciNet.rdf#rciNet/Domain/vlan/7630992e-ac99-4a78-beee-1b504b29cef0/vlan;p_rmg=orca.manage.beans.LeaseReservationMng@5c65d200 m_p 1

SETTING num.parent.new to 1
```
in the invalid case it doesn't happen
```
m_r_map for uri http://geni-orca.renci.org/owl/rciNet.rdf#rciNet/Domain/vlan/20dc6610-8fa2-4c52-9cdc-c72325ac39b6/vlan returned null
contains test false

m_r_map for uri http://geni-orca.renci.org/owl/rciNet.rdf#rciNet/Domain/vlan/a0cf1e86-1a3c-4ac8-8918-a0d0fd3c5a03/vlan returned null
contains test false
```
The precededBy set is modified by the function that was combined by Alan from ModifyHandler into CloudHandler.createInterface(). So at least it seems this problem can be traced to the previous bug fix. More investigation needed.

**Comment by ibaldin** [comment]
The fix above removes the early exit out of the loop for CloudHandler.setEdgeNeighbourhood():

```
                    for (Object de : ce.getDependencies().toArray()) {
                        if (de instanceof NetworkConnection) {
                        	System.out.println("NetworkConnection");
                            NetworkConnection ne = (NetworkConnection) de;
                            // this IF doesn't fire sometimes for #180
                            if (ne.getName().equals(ncByInterface.getName())) {
                                if (ne.getFirstConnectionElement() != null)
                                    link_device = (DomainElement) ne.getFirstConnectionElement();
                                System.out.println("LOOP");
                                link_device.setFollowedBy(edge_device, new_intf.getResource());
                                edge_device.setPrecededBy(link_device, new_intf.getResource());
                            }
                            //break;
                        }
```
Not clear what the implications are for the code overall. Because the order of traversal of elements in the for loop is random, sometimes the correct network element is identified and the loop is exited and sometimes and incorrect one (and in this one because of the break statement it never gets to the correct one). Need more consultation with @YufengXin .

**Comment by ibaldin** [comment]
This is related to #178, the fix for 178 produced this problem. 178 will remain closed.

**Comment by ibaldin** [comment]
Per @YufengXin the issue may be due to merging of createInterface() functions between CloudHandler and ModifyHandler, where the existence of extra elements on the list wasn't anticipated (for the create case). He is looking into it. Currently the code passes all existing unit tests.

**Comment by YufengXin** [comment]
Just wondering if the problem only happens when delete than add, consecutively adding nodes works fine?

**Comment by ibaldin** [comment]
It happens within the same unit test with probability of about 70% failure 30% success. The unit test does not vary what it does. 

As I mentioned the problem itself is resolved (by commenting out the break statement you see above). The issue is - is it the right fix for all situations.

**Comment by YufengXin** [comment]
I did some more tests in both unit tests and the emulation tests and didn't see failures. This issue can be closed by this fix.

The "break" statement should be inside the "IF" block, ie, as long as the connection is found, no need to further loop. But removing "break" statement works just fine.

**Comment by ibaldin** [comment]
Resolved in #185
